rust/src/applayer.rs

```diff
@@ -535,14 +535,17 @@ pub trait Transaction {
 }
 
 pub trait State<Tx: Transaction> {
-    fn get_transactions(&self) -> &[Tx];
+    /// Return the number of transactions in the state's transaction collection.
+    fn get_transaction_count(&self) -> usize;
+
+    /// Return a transaction by its index in the container.
+    fn get_transaction_by_index(&self, index: usize) -> Option<&Tx>;
 
     fn get_transaction_iterator(&self, min_tx_id: u64, state: &mut u64) -> AppLayerGetTxIterTuple {
         let mut index = *state as usize;
-        let transactions = self.get_transactions();
-        let len = transactions.len();
+        let len = self.get_transaction_count();
         while index < len {
-            let tx = &transactions[index];
+            let tx = self.get_transaction_by_index(index).unwrap();
```

Comment:
Should we rather have an iterator ?

Reply:
I don't think we can make it generic, at least not without a bunch more work.. The iterator returned from VecDeque and Vec are different types, and I couldn't find away to be generic over both of them.

I did try using `.iter()` and `.iter().enumerate()` on a local DNS transaction iterator for a purely idiomatic Rust approach expecting a tiny noticeable performance different.  `.enumerate()` to get the index from the iterator instead of manually updating it was about 1000 times slower than using the index approach.  And I didn't notice a performance difference between an `.iter()` and manually updating the index var, or using the purely indexed approach here.  I was hoping to, to justify the extra work, but I can't justify it, and the added complexity right now.